### PR TITLE
Version 14.1.5 fixes, on the way to v15.

### PR DESCRIPTION
**Pull request by EBatTiVo**

### CODE REVIEW

File: src/common/com/intellij/plugins/haxe/ide/formatter/HaxeFormattingModel.java [hunk 1]
```diff
@@ -32,10 +35,18 @@
...
+
   public HaxeFormattingModel(final PsiFile file,
                              CodeStyleSettings settings,
                              final Block rootBlock) {
     myModel = FormattingModelProvider.createFormattingModelForPsiFile(file, rootBlock, settings);
+
+    shiftIndentInsideRangeInvoker = new ApiInvoker<TextRange>(myModel, "shiftIndentInsideRange",
+                                                              IdeaTarget.IS_VERSION_14_1_5_COMPATIBLE
+                                                              ? new Class<?>[] {ASTNode.class, TextRange.class, int.class }
+                                                              : new Class<?>[] {TextRange.class, int.class} );
```
Comment (as3boyan) [inline src/common/com/intellij/plugins/haxe/ide/formatter/HaxeFormattingModel.java]: It is so complicated, I am afraid I did a similar job in a recent commit

**Comment by EBatTiVo** [comment]
I see that you did.  I was trying to get done before you got back to work this morning.  Oh well.  

At any rate, this is a bit complicated, but it actually makes the changes easier; the files can stay in the original places.  The ApiInvoker class does most of the magic, and it doesn't have to be repeated.  Nor do the extra changes to the constants.  We can come up with a more readable pattern for setting up the invoker, if you don't like the one-liner that I have there.  It can easily be

if (IS_VERSION_X_COMPATIBLE) {
  myInvoker = new ApiInvoker<return_type>(this, "apiName", arg1type.class, arg2type.class);
} else {
  myInvoker = new ApiInvoker<return_type>(this, "apiName", arg1type.class);
}

**Comment by as3boyan** [comment]
In Haxe writing conditional code is so much easier

**Comment by as3boyan** [comment]
>I see that you did. I was trying to get done before you got back to work this morning. Oh well. 

I am too fast when it comes to do double work :smiley:

**Comment by as3boyan** [comment]
I like the competitive spirit of this race) Anyway, it is good, to see various solutions) :+1:

**Comment by EBatTiVo** [comment]
@as3boyan I've got this tested, now and it works very well.  I think that we should take this PR, rather than the one you've posted -- mostly because it will allow us to not have another set of duplicate files to maintain, and we can merge most of those older ones back in.  I'm going to steal some of your work, though.

**Comment by as3boyan** [comment]
Ok, no problem, we have a enough things to do

**Comment by sganapavarapu1** [comment]
looks good to me, approved

File: src/common/com/intellij/plugins/haxe/build/ApiInvoker.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+      // naturally.
+      myMethodToInvoke = klass.getDeclaredMethod(methodName, argTypes);
+    } catch (NoSuchMethodException e) {
+      throw new UnsupportedApiException("Didn't find " + myDebugName + ";", e);
+    }
+  }
+
+  public R invoke(Object... args) {
+    try {
+      return (R) myMethodToInvoke.invoke(myInstance, args);
```
Comment (as3boyan) [inline src/common/com/intellij/plugins/haxe/build/ApiInvoker.java]: Dynamic call?

Comment (EBatTiVo) [inline src/common/com/intellij/plugins/haxe/build/ApiInvoker.java]: Yes, it uses reflection to get around the fact that the compiler wants to see correct syntax in all branches.  This way, we can look up the function and bind it at run-time.  It gets us away from needing two source files, at the small cost of run-time linking (and possibly failure) and the potential loss of some optimizations.

**Pull request merged**